feat: add CONTRIBUTING.md file (#11)
faast-rt/lambdo: CONTRIBUTING.md

# Contributing to FAAST

> Thanks for your interest in improving the project! This document provides a step-by-step guide for general contributions to FAAST.

Reviewer: ```suggestion
Thanks for your interest in improving the project! This document provides a guideline for general contributions to FAAST.
```
This is not a step by step guide. We just give the rules to follow.

## Getting started

There are there general ways you can contribute to this repo:

- Proposing an enhancement or new feature
- Reporting a bug or regression
- Contributing changes to the source code

For the first two, refer to the [GitHub Issues](https://github.com/faast-rt/lambdo/issues/new/choose) which guides you through the available options along with the needed information to collect.

> > To setup your development environment, please refer to the [README](./README.md).

Reviewer: Not sure why this is here. Would make more sense in "Contributing changes"

Reviewer: (Comment applies to the last line of the suggestion)

## Contributing changes

> _Prior to opening a pull request, it is recommended to open an issue first to ensure the maintainers can review intended changes. Exceptions to this rule include fixing non-functional source such as code comments, documentation or other supporting files._

Reviewer: We might be able to slightly change the wording to make the message clearer:

```suggestion
_Please open an issue before adding a new feature or making major changes to existing one. It allows the community to discuss the relevancy of the issue, as well as the caveats one should avoid when implementing it._
```
(Far from perfect, but gets the point across more accurately)

> _Prior to opening a pull request, it is recommended to open an issue first to ensure the maintainers can review intended changes. Exceptions to this rule include fixing non-functional source such as code comments, documentation or other supporting files._

Reviewer 2: Sounds good

> _Prior to opening a pull request, it is recommended to open an issue first to ensure the maintainers can review intended changes. Exceptions to this rule include fixing non-functional source such as code comments, documentation or other supporting files._

Reviewer 3: I quite liked the "Exceptions to this rule include fixing non-functional source such as code comments, documentation or other supporting files." part

Proposing source code changes is done through [GitHub's standard pull request workflow](https://docs.github.com/en/get-started/quickstart/github-flow#create-a-pull-request).

If your branch is a work-in-progress then please start by creating your pull requests as draft, by clicking the down-arrow next to the `Create pull request` button and instead selecting `Create draft pull request`.

This will defer the automatic process of requesting a review from the FAAST team and significantly reduces noise until you are ready. Once you are happy, you can click the `Ready for review` button.

> 

Reviewer: We don't talk where the contributor should contribute. I think they should fork the repo and create PRs from their repo. Adding a small sentence about that seems relevant.

### Pull Request Guidelines

A good pull request includes:

> - A high-level description of the changes, including links to any issues that are related by adding comments like `Resolves #NNN` to your description. See [Linking a Pull Request to an Issue](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) for more information.

Reviewer: Does it ? The high level description should be in the issue, not in the PR, right ? PR should discuss the implementation.

Reviewer 3: Yes it does, you don't necessarily want to check all the issues to understand what a PR is about IMO

Author: For us, it does too.

- Unit tests where appropriate. Bug fixes will benefit from the addition of regression tests. New features will not be accepted without suitable test coverage!
- No more commits than necessary. Sometimes having multiple commits is useful for telling a story or isolating changes from one another, but please squash down any unnecessary commits that may just be for clean-up, comments or small changes.
- No additional external dependencies that aren't absolutely essential. Please do everything you can to avoid pulling in additional libraries/dependencies into `Cargo.toml` as we will be very critical of these.

### Commits conventions

Our commit convention follow the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)

To be more explicit, we will describe below the commit conventions message.

#### Commit Message Format

Each commit message consists of a header, a body and a footer. The header has a special format that includes a type, a scope and a subject:

```
<type>[optional scope]: <description>

[optional body]

[optional footer(s)]
```

> Any line of the commit message cannot be longer 100 characters! This allows the message to be easier to read on GitHub as well as in various git tools.

Reviewer 3: ```suggestion
Any line of the commit message cannot be longer than 100 characters! This allows the message to be easier to read on GitHub as well as in various git tools.
```

- use the imperative, present tense: `change` not `changed` nor `changes`
- don't capitalize the first letter
- no dot (.) at the end

#### Body

Just as in the subject, use the imperative, present tense: `change` not `changed` nor `changes`. The body should include the motivation for the change and contrast this with previous behavior.

#### Footer

The footer should contain any information about Breaking Changes and is also the place to [reference GitHub issues that the commit closes](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

**Breaking Changes** should start with the word `BREAKING CHANGE:` with a space or two newlines. The rest of the commit message is then used for this.

#### Sign-off

Please consider signing the commit message at least with `Signed-Off-By`. This is a way to certify that you wrote the code or otherwise have the right to pass it on as an open-source patch. The process is simple: if you can certify the [Developer's Certificate of Origin](https://developercertificate.org/) (DCO).

To perform a sign-off with `git`, use :

```bash
git commit -s #(or --signoff)
```

## Get help

> If you have questions about the contribution process, please start a [GitHub discussion](https://github.com/faast-rt/lambdo/discussions) or send your question to the contact email address: [contact@faast-rt.com](mailto://contact@faast-rt.com).

Reviewer: Do we really need all this text about conventional commits ? This is 2/3 of our contributing. We can have a small bullet point list to recap what it is about, but the reader is expected to read the specs if they don't know it.

Reviewer 4: I prefer to have something that is clearly defined than a reference to another convention that can change at any time without realizing it.
